local/include/core/http3.h

```diff
@@ -0,0 +1,401 @@
+};
+} // namespace bwf
+
+BufferWriter &bwformat(BufferWriter &w, bwf::Spec const &spec, bwf::Ngtcp2Error const &error);
+BufferWriter &bwformat(BufferWriter &w, bwf::Spec const &spec, bwf::Nghttp3Error const &error);
+} // namespace SWOC_VERSION_NS
+} // namespace swoc
+
+/** Encapsulate the buffer for the QUIC hanshake. */
+class QuicHandshake
+{
+public:
+  QuicHandshake();
+  ~QuicHandshake() = default;
+
+  QuicHandshake(QuicHandshake const &) = delete;
+  QuicHandshake &operator=(QuicHandshake const &) = delete;
+
+public:
+  /** This is the maximum number of bytes we expect to use for the QUIC
+   * handshake.
+   *
+   * This max value is taken from CURL code which has a comment expressing
+   * tentative hope that this should be large enough. There is an assertion in
+   * our (and theirs, I believe) implementation guarding this invariant. If we
+   * trip that, we may need to expand this.
+   */
+  static constexpr size_t max_handshake_size = 4 * 1024;
+
+  /** This contains the storage for the QUIC handshake. */
+  std::vector<char> buf;
+};
+
+/** The various elements related to the ngtcp2 API calls.
+ *
+ * For reference, this is based off of curl's struct quicsocket in ngtcp2.h.
+ */
+class QuicSocket
+{
+public:
+  QuicSocket();
+  ~QuicSocket();
+  QuicSocket(QuicSocket const &) = delete;
+  QuicSocket &operator=(QuicSocket const &) = delete;
+
+  /** Open a QUIC log file for writing.
+   *
+   * This assumes that scid has been previously configured.
+   */
+  swoc::Errata open_qlog_file();
+
+  /** Randomly populate an array of a given size.
+   *
+   * This is used to initialize the various connection ids.
+   *
+   * @param[in] array The buffer to populate
+   *
+   * @param[in] array_len The number of bytes in the array to populate.
+   */
+  static void randomly_populate_array(uint8_t *array, size_t array_len);
+
+  /** Configure QUIC logging for the provided directory.
+   *
+   * @param[in] qlog_dir The directory into which QUIC log files should be
+   * written.
+   */
+  static swoc::Errata configure_qlog_dir(swoc::TextView qlog_dir);
+
+  /** The callback function for ngtcp2 QUIC logging.
+   *
+   * For details, see the ngtcp2 document for ngtcp2_qlog_write.
+   */
+  static void qlog_callback(void *user_data, uint32_t flags, const void *data, size_t datalen);
+
+public:
+  ngtcp2_conn *qconn = nullptr;
+  ngtcp2_cid dcid;
+  ngtcp2_cid scid;
+  uint32_t version = 0;
+  ngtcp2_settings settings;
+  ngtcp2_transport_params transport_params;
+  SSL_CTX *sslctx = nullptr;
+  SSL *ssl = nullptr;
+  /// 3 is the maximum enum value in ngtcp2_crypto_level.
+  static constexpr int MAX_NGTCP2_CRYPTO_LEVEL = 3;
+  // The indexing starts with 0, thus if the MAX_NGTCP2_CRYPTO_LEVEL is 3,
+  // there can be 4 entries (0 to 3, inclusive).
+  QuicHandshake crypto_data[MAX_NGTCP2_CRYPTO_LEVEL + 1];
+  /* The last TLS alert description generated by the local endpoint */
+  uint8_t tls_alert = 0;
+  struct sockaddr_storage local_addr;
```

Comment:
I'd be tempted to use `IPEndpoint` here instead. You don't really need `sockaddr_storage`, just IPv6 capable.